### PR TITLE
Port master_append_table_to_shard to new connection API

### PR DESCRIPTION
**Pull request by lithp**
Part of #1101.

Ported it over to use the transaction management stuff as well, so we can fearlessly cancel halfway through it, I believe.

### CODE REVIEW

File: src/backend/distributed/master/master_stage_protocol.c [hunk 1]
```diff
@@ -261,67 +258,41 @@ master_append_table_to_shard(PG_FUNCTION_ARGS)
...
 	foreach(shardPlacementCell, shardPlacementList)
 	{
 		ShardPlacement *shardPlacement = (ShardPlacement *) lfirst(shardPlacementCell);
-		char *workerName = shardPlacement->nodeName;
-		uint32 workerPort = shardPlacement->nodePort;
-		List *queryResultList = NIL;
+		MultiConnection *connection = GetPlacementConnection(FOR_DML, shardPlacement,
+															 NULL);
+		PGresult *queryResult = NULL;
+		int executeCommand = 0;
```
Comment (marcocitus) [inline src/backend/distributed/master/master_stage_protocol.c]: Maybe executeResult

File: src/backend/distributed/master/master_stage_protocol.c [hunk 2]
```diff
@@ -261,67 +258,41 @@ master_append_table_to_shard(PG_FUNCTION_ARGS)
...
-		UpdateShardPlacementState(placementId, FILE_INACTIVE);
-
-		ereport(WARNING, (errmsg("could not append table to shard \"%s\" on node "
-								 "\"%s:%u\"", shardQualifiedName, workerName,
-								 workerPort),
-						  errdetail("Marking this shard placement as inactive")));
-	}
-
 	/* update shard statistics and get new shard size */
 	newShardSize = UpdateShardStatistics(shardId);
```
Comment (marcocitus) [inline src/backend/distributed/master/master_stage_protocol.c]: Hm, we probably want to error out before we do more work like getting statistics. Maybe count successes.

File: src/include/distributed/transaction_management.h [hunk 3]
```diff
@@ -66,6 +66,9 @@ extern CoordinatedTransactionState CurrentCoordinatedTransactionState;
 /* list of connections that are part of the current coordinated transaction */
 extern dlist_head InProgressTransactions;
 
+/* whether we've been asked to use 2PC (by calling CoordinatedTransactionUse2PC()) */
+extern bool CoordinatedTransactionUses2PC;
```
Comment (anarazel) [inline src/include/distributed/transaction_management.h]: Uh, wait, what?  I don't think this is is a good idea.   If we want this, we should have a wrapper for CheckForFailedPlacements that's callable externaly.

**Comment by lithp** [comment]
I like the idea of a new function, not only is the signature strange but
the error messages weren't really written to appear during a statement.
I'll open a ticket to fix it in 6.2?

File: src/include/distributed/transaction_management.h (hunk 3 above)
Comment (jasonmp85) [inline src/include/distributed/transaction_management.h]: Yeah I'm not too familiar with the Transaction API, but this kind of global state that needs setting and unsetting during methods seems like a bad pattern or a leaky abstraction.

**Comment by sumedhpathak** [comment]
@lithp Could you then follow-up and open a new issue? What is your and @anarazel 's estimate for cleaning up this API? Looking to see impact on 6.2 prioritization.

**Comment by lithp** [comment]
@sumedhpathak This is not a big change, a day at most.

File: src/include/distributed/transaction_management.h (hunk 3 above)
Comment (lithp) [inline src/include/distributed/transaction_management.h]: I've opened a PR: https://github.com/citusdata/citus/pull/1169

**Pull request merged**